File: ts/src/storage/page_cache.ts
// LRU page cache — keeps hot pages in memory, evicts cold ones

import { LRUCache } from "../util/lru.ts";
import { Page } from "./page.ts";
import { PAGE_SIZE, FILE_HEADER_SIZE } from "../types.ts";

export class PageCache {
  private cache: LRUCache<number, Page>;
  private dirtyPages = new Set<number>();
  private file: Deno.FsFile;

  constructor(file: Deno.FsFile, maxPages: number = 1024) {
    this.file = file;
    this.cache = new LRUCache<number, Page>(maxPages, (pageNumber, page) => {
      // On eviction, flush if dirty
      if (this.dirtyPages.has(pageNumber)) {
        this.flushPageSync(pageNumber, page);
      }
    });
  }

  async getPage(pageNumber: number): Promise<Page> {
    const cached = this.cache.get(pageNumber);
    if (cached) return cached;

    // Read from disk
    const page = await this.readPageFromDisk(pageNumber);
    this.cache.set(pageNumber, page);
    return page;
  }

  putPage(pageNumber: number, page: Page): void {
    this.cache.set(pageNumber, page);
  }

  markDirty(pageNumber: number): void {
    this.dirtyPages.add(pageNumber);
  }

  async flushAll(): Promise<void> {
    for (const pageNumber of this.dirtyPages) {
      const page = this.cache.get(pageNumber);
      if (page) {
        await this.writePageToDisk(pageNumber, page);
      }
    }
    this.dirtyPages.clear();
    await this.file.sync();
  }

  async flushPage(pageNumber: number): Promise<void> {
    if (!this.dirtyPages.has(pageNumber)) return;
    const page = this.cache.get(pageNumber);
    if (page) {
      await this.writePageToDisk(pageNumber, page);
      this.dirtyPages.delete(pageNumber);
    }
  }

  private async readPageFromDisk(pageNumber: number): Promise<Page> {
    const offset = FILE_HEADER_SIZE + pageNumber * PAGE_SIZE;
    const buf = new Uint8Array(PAGE_SIZE);
    await this.file.seek(offset, Deno.SeekMode.Start);
    await readFull(this.file, buf);
    return new Page(buf);
  }

  private async writePageToDisk(pageNumber: number, page: Page): Promise<void> {
    const offset = FILE_HEADER_SIZE + pageNumber * PAGE_SIZE;
    await this.file.seek(offset, Deno.SeekMode.Start);
    await writeFull(this.file, page.data);
  }

  private flushPageSync(pageNumber: number, page: Page): void {
    // Synchronous flush during eviction — use blocking write
    const offset = FILE_HEADER_SIZE + pageNumber * PAGE_SIZE;
    this.file.seekSync(offset, Deno.SeekMode.Start);
    this.file.writeSync(page.data);
    this.dirtyPages.delete(pageNumber);
  }

  isDirty(pageNumber: number): boolean {
    return this.dirtyPages.has(pageNumber);
  }

  get dirtyCount(): number {
    return this.dirtyPages.size;
  }

  clear(): void {
    this.dirtyPages.clear();
    this.cache.clear();
  }
}

async function readFull(file: Deno.FsFile, buf: Uint8Array): Promise<void> {
  let offset = 0;
  while (offset < buf.byteLength) {
    const n = await file.read(buf.subarray(offset));
    if (n === null) break;
    offset += n;
  }
}

async function writeFull(file: Deno.FsFile, buf: Uint8Array): Promise<void> {
  let offset = 0;
  while (offset < buf.byteLength) {
    const n = await file.write(buf.subarray(offset));
    offset += n;
  }
}

File: internal/storage/page_cache.go
package storage

import (
	"os"
	"sync"

	"github.com/marcisbee/flop/internal/schema"
	"github.com/marcisbee/flop/internal/util"
)

// PageCache is an LRU cache for pages with dirty tracking.
type PageCache struct {
	mu         sync.Mutex
	cache      *util.LRUCache[uint32, *Page]
	dirtyPages map[uint32]bool
	file       *os.File
}

func NewPageCache(file *os.File, maxPages int) *PageCache {
	pc := &PageCache{
		dirtyPages: make(map[uint32]bool),
		file:       file,
	}
	lru := util.NewLRUCache[uint32, *Page](maxPages)
	lru.OnEvict = func(pageNum uint32, page *Page) error {
		if pc.dirtyPages[pageNum] {
			return pc.flushPageSync(pageNum, page)
		}
		return nil
	}
	pc.cache = lru
	return pc
}

func (pc *PageCache) GetPage(pageNumber uint32) (*Page, error) {
	pc.mu.Lock()
	defer pc.mu.Unlock()

	if p, ok := pc.cache.Get(pageNumber); ok {
		return p, nil
	}

	p, err := pc.readPageFromDisk(pageNumber)
	if err != nil {
		return nil, err
	}
	if err := pc.cache.Set(pageNumber, p); err != nil {
		return nil, err
	}
	return p, nil
}

func (pc *PageCache) PutPage(pageNumber uint32, page *Page) error {
	pc.mu.Lock()
	defer pc.mu.Unlock()
	return pc.cache.Set(pageNumber, page)
}

func (pc *PageCache) MarkDirty(pageNumber uint32) {
	pc.mu.Lock()
	defer pc.mu.Unlock()
	pc.dirtyPages[pageNumber] = true
}

func (pc *PageCache) FlushAll() error {
	pc.mu.Lock()
	defer pc.mu.Unlock()

	for pageNum := range pc.dirtyPages {
		if p, ok := pc.cache.Get(pageNum); ok {
			if err := pc.writePageToDisk(pageNum, p); err != nil {
				return err
			}
		}
	}
	pc.dirtyPages = make(map[uint32]bool)
	return pc.file.Sync()
}

func (pc *PageCache) readPageFromDisk(pageNumber uint32) (*Page, error) {
	offset := int64(schema.FileHeaderSize) + int64(pageNumber)*int64(schema.PageSize)
	buf := make([]byte, schema.PageSize)
	n, err := pc.file.ReadAt(buf, offset)
	if err != nil && n < schema.PageSize {
		return nil, err
	}
	page := NewPage(buf)
	if err := page.Validate(pageNumber); err != nil {
		return nil, err
	}
	return page, nil
}

func (pc *PageCache) writePageToDisk(pageNumber uint32, page *Page) error {
	offset := int64(schema.FileHeaderSize) + int64(pageNumber)*int64(schema.PageSize)
	data := page.Snapshot()
	_, err := pc.file.WriteAt(data[:], offset)
	return err
}

func (pc *PageCache) flushPageSync(pageNumber uint32, page *Page) error {
	offset := int64(schema.FileHeaderSize) + int64(pageNumber)*int64(schema.PageSize)
	data := page.Snapshot()
	if _, err := pc.file.WriteAt(data[:], offset); err != nil {
		return err
	}
	delete(pc.dirtyPages, pageNumber)
	return nil
}

func (pc *PageCache) IsDirty(pageNumber uint32) bool {
	pc.mu.Lock()
	defer pc.mu.Unlock()
	return pc.dirtyPages[pageNumber]
}

func (pc *PageCache) DirtyCount() int {
	pc.mu.Lock()
	defer pc.mu.Unlock()
	return len(pc.dirtyPages)
}

func (pc *PageCache) Clear() {
	pc.mu.Lock()
	defer pc.mu.Unlock()
	pc.dirtyPages = make(map[uint32]bool)
	pc.cache.Clear()
}
